File: src/features/providers/builtin/deepseek.ts
// ABOUTME: DeepSeek OpenAI-compatible plugin with first-class thinking policy.
// ABOUTME: Reuses chat.completions wire format; owns detect thinking defaults.
import type { AuthSchemeV1, CredentialKind } from "../../../storage/types";
import type {
  ChatBuildInput,
  DetectPolicy,
  DetectPolicyInput,
  ModelListBuildInput,
  ProviderHttpResponse,
  ProviderPlugin,
  ProviderPluginManifest,
  ProviderWireRequest,
  SseEvent,
  StreamParseResult,
} from "../types";
import { ProviderProtocolError } from "../types";
import {
  buildOpenAiChatCompletions,
  parseOpenAiChatContent,
  parseOpenAiPage,
  parseOpenAiStreamDelta,
} from "./openaiShared";

const DETECT_MAX_TOKENS_THINKING = 2048;

const MANIFEST: ProviderPluginManifest = {
  id: "deepseek",
  label: "DeepSeek",
  defaultBaseUrl: "https://api.deepseek.com",
  supportedCredentialKinds: ["none", "api_key", "bearer"],
  capabilities: {
    modelListing: true,
    streaming: true,
    textGeneration: true,
    imageInput: true,
  },
};

export const deepseekPlugin: ProviderPlugin = {
  manifest: MANIFEST,

  resolveAuthScheme(credentialKind: CredentialKind): AuthSchemeV1 {
    if (credentialKind === "none") {
      return { schemaVersion: 1, type: "none" };
    }
    return { schemaVersion: 1, type: "bearer" };
  },

  buildModelListRequest(input: ModelListBuildInput): ProviderWireRequest {
    void input;
    return {
      method: "GET",
      relativePath: "models",
      query: [],
      headers: {},
      body: null,
    };
  },

  parseModelListPage(response: ProviderHttpResponse) {
    return parseOpenAiPage(response);
  },

  buildChatRequest(input: ChatBuildInput): ProviderWireRequest {
    const wire = buildOpenAiChatCompletions({
      modelKey: input.modelKey,
      systemPrompt: input.systemPrompt,
      userPrompt: input.userPrompt,
      temperature: input.temperature,
      maxTokens: input.maxTokens,
      imagePngBase64: input.imagePngBase64,
      stream: input.stream,
    });
    if (input.thinking != null && wire.body) {
      const payload = JSON.parse(wire.body) as Record<string, unknown>;
      payload.thinking = { type: input.thinking ? "enabled" : "disabled" };
      return { ...wire, body: JSON.stringify(payload) };
    }
    return wire;
  },

  parseChatResponse(response: ProviderHttpResponse): string {
    return parseOpenAiChatContent(response);
  },

  parseStreamEvent(event: SseEvent): StreamParseResult {
    if (!event.data || event.data === "[DONE]") {
      return { kind: "ignore" };
    }
    let value: unknown;
    try {
      value = JSON.parse(event.data) as unknown;
    } catch {
      throw new ProviderProtocolError("stream event is not JSON");
    }
    const delta = parseOpenAiStreamDelta(value);
    if (delta == null) {
      return { kind: "ignore" };
    }
    return { kind: "delta", text: delta };
  },

  getDetectPolicy(input: DetectPolicyInput): DetectPolicy {
    void input;
    return { thinking: false, maxTokens: DETECT_MAX_TOKENS_THINKING };
  },
};

File: src-tauri/src/adapters/builtin/deepseek.rs
// ABOUTME: DeepSeek OpenAI-compatible adapter with first-class thinking policy.
// ABOUTME: Reuses chat.completions wire format; owns detect/translate thinking defaults.
use crate::adapters::builtin::openai_shared::{
  build_openai_chat_completions, parse_openai_chat_content, parse_openai_page, parse_openai_stream_delta,
};
use crate::adapters::protocol::{AdapterMeta, DETECT_MAX_TOKENS_THINKING, DetectChatPolicy, ProviderAdapter};
use crate::adapters::transport::{AuthApplication, ChatCompletionRequest, ParsedPage, TransportError};
use crate::domain::provider::CredentialKind;

/// DeepSeek API (OpenAI-compatible wire format + thinking controls).
///
/// DeepSeek V4 defaults thinking to **enabled**. Detection turns thinking off and
/// raises `max_tokens` so a short answer still fits if a gateway ignores the toggle.
#[derive(Debug, Default, Clone, Copy)]
pub struct DeepSeekAdapter;

impl ProviderAdapter for DeepSeekAdapter {
  fn meta(&self) -> AdapterMeta {
    AdapterMeta {
      id: "deepseek",
      label: "DeepSeek",
      default_base_url: Some("https://api.deepseek.com"),
    }
  }

  fn secret_required(&self, credential_kind: CredentialKind) -> bool {
    matches!(credential_kind, CredentialKind::ApiKey | CredentialKind::Bearer)
  }

  fn auth_application(&self, credential_kind: CredentialKind) -> Result<AuthApplication, TransportError> {
    match credential_kind {
      CredentialKind::None => Ok(AuthApplication::None),
      CredentialKind::ApiKey | CredentialKind::Bearer => Ok(AuthApplication::BearerHeader),
    }
  }

  fn models_path(&self) -> &'static str {
    "models"
  }

  fn parse_models_page(&self, value: &serde_json::Value) -> Result<ParsedPage, TransportError> {
    parse_openai_page(value)
  }

  fn build_chat(
    &self,
    request: &ChatCompletionRequest,
    stream: bool,
  ) -> Result<(url::Url, serde_json::Value), TransportError> {
    let (url, mut payload) = build_openai_chat_completions(
      &request.base_url,
      &request.model_key,
      &request.system_prompt,
      &request.user_prompt,
      request.temperature,
      request.max_tokens,
      request.image_png_base64.as_deref(),
      stream,
    )?;
    apply_deepseek_thinking(&mut payload, request.thinking);
    Ok((url, payload))
  }

  fn parse_chat_content(&self, value: &serde_json::Value) -> Result<String, TransportError> {
    parse_openai_chat_content(value)
  }

  fn parse_stream_delta(&self, _event_name: Option<&str>, value: &serde_json::Value) -> Option<String> {
    parse_openai_stream_delta(value)
  }

  fn detect_chat_policy(&self, _model_key: &str, _base_url: &str) -> DetectChatPolicy {
    DetectChatPolicy {
      // Best-effort disable; some relays ignore the toggle and still emit CoT.
      thinking: Some(false),
      max_tokens: DETECT_MAX_TOKENS_THINKING,
    }
  }
}

/// DeepSeek-only `thinking.type` control (`enabled` / `disabled`).
fn apply_deepseek_thinking(payload: &mut serde_json::Value, thinking: Option<bool>) {
  if let Some(enabled) = thinking {
    payload["thinking"] = serde_json::json!({
      "type": if enabled { "enabled" } else { "disabled" }
    });
  }
}
